keps/407-gang-scheduling/README.md

```diff
@@ -0,0 +1,511 @@
+		// no special handling is needed for worker pods
+		...
+	}
+
+	// Set the podgroup affiliation for the leader pod and worker pod
+	if utilfeature.DefaultFeatureGate.Enabled(features.PodGroupPerReplica) {
+		err = p.SchedulerProvider.InjectPodGroupMetadata(pod)
+		if err != nil {
+			return err
+		}
+	}
+
+	...
+}
+```
+Because the name of podgroup is `lws.Name-groupIndex`, groupIndex can only be obtained after the leader pod is created,
+and then the annotation or label of the leader pod can be patched to set the podgroup it belongs to.
+
+While the worker pods have already been set with the groupIndex when the pod controller creates the worker sts,
+so we can directly get the setting of `leaderworkerset.sigs.k8s.io/group-index` from the worker pod labels to set the podgroup name.
+### Pod Controller
+```go
+func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	...
+	// Create headless service per replica
+	if leaderWorkerSet.Spec.NetworkConfig != nil && *leaderWorkerSet.Spec.NetworkConfig.SubdomainPolicy == leaderworkerset.SubdomainUniquePerReplica {
+		if err := controllerutils.CreateHeadlessServiceIfNotExists(ctx, r.Client, r.Scheme, &leaderWorkerSet, pod.Name, map[string]string{leaderworkerset.SetNameLabelKey: leaderWorkerSet.Name, leaderworkerset.GroupIndexLabelKey: pod.Labels[leaderworkerset.GroupIndexLabelKey]}, &pod); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Create podgroup per replica
+	if utilfeature.DefaultFeatureGate.Enabled(features.PodGroupPerReplica) {
+		err = r.SchedulerProvider.CreatePodGroupIfNotExists(ctx, &leaderWorkerSet, &pod)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	...
+}
+```
+In the pod controller, when the pod controller watched the creation of a Leader Pod, it will create a podgroup corresponding to the groupIndex, 
+set the required resources and the minimum number of pods that should be Running (minAvailable), 
+and set the ownerReference of the podgroup to the leader pod. In this way, the lifecycle of the podgroup is the same as the leader pod. 
+When an update occurs to lws, the leader pod will be recreated, and the podgroup should also be recreated. 
+During scaling down/deletion, the podgroup should also be deleted. During rolling updates, if the user has configured maxSurge, 
+then additional leader pods should be created, and corresponding additional podgroups should also be created to schedule these additional pods, 
+so that lws can perform rolling updates successfully.
+
+### StartupPolicy Handling
+
+Different `StartupPolicy` should have different MinMember settings:
+- **LeaderCreated Policy (Default):** MinMember is set to the full replica size (1 leader + (size-1) workers). All pods in the replica must be scheduled together. 
+- **LeaderReady Policy:** MinMember is set to 1, therefore leader pod can be scheduled immediately and won't be blocked by gang scheduler.
```

Comment:
so one risk here is once the podGroup doesn't support both minMember and minResources, it can not match with startupPolicy perfectly. Let's add it as a risk.